include/rmm/mr/device_memory_resource.hpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cstddef>
+
+// forward decl
+using cudaStream_t = struct CUstream_st*;
+
+namespace rmm {
+namespace mr {
+/**---------------------------------------------------------------------------*
+ * @brief Base class for all libcudf device memory allocation.
+ *
+ * This class serves as the interface that all custom device memory
+ * implementations must satisfy.
+ *
+ * There are two private, pure virtual functions: `do_allocate` and
+ * `do_deallocate` that must be implemented. Optionally, the `is_equal` virtual
+ * function may be overriden from the default of simply performing an identity
+ * comparison.
+ *
+ * The public, non-virtual functions `allocate`, `deallocate`, and `is_equal`
+ * simply call the private virtual functions. The reason for this is to allow
+ * implementing shared, default behavior in the base class. For example, the
+ * base class' `allocate` function may log every allocation, no matter what
+ * derived class implementation is used.
+ *
+ *---------------------------------------------------------------------------**/
+class device_memory_resource {
+ public:
+  device_memory_resource() = default;
+  virtual ~device_memory_resource() = default;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer will have at minimum 256 byte alignment.
```

Comment:
Maybe add an assert checking for this alignment on the pointer returned by `do_allocate`?